File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        brackets = {
            ')':'(',
            ']':'[',
            '}':'{'
        }
        stack = []
        
        for i in s:
            if i in  ['(','[','{']:
                stack.append(i)
            else:
                if stack:
                    if brackets[i] == stack[-1]:
                        stack.pop()
                        continue
                    else:
                        return False
                else:
                    return False
        return not stack
                
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        # stack=[]
        # # end=[")","}","]"]
        # # start=["(","[","{"]
        # parent={"(":")","[":"]","{":"}"}
        # for par in s:
        #     if par not in parent.values():
        #         stack.append(par)
        #     else:
        #         if stack:
        #             val=stack.pop()
        #             if par==parent[val]:
        #                 continue
        #             else:
        #                 return False
        #         else:
        #             return False
        # if stack:
        #     return False
        # else:
        #     return True
                
                
            

File: 0020-valid-parentheses/0020-valid-parentheses.java
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> bracket = new HashMap<>();
        bracket.put(')', '(');
        bracket.put(']', '[');
        bracket.put('}', '{');
        Stack<Character> st = new Stack<>();

        for (char c : s.toCharArray()) {
            if (!bracket.containsKey(c)) {
                st.push(c);  // Use push instead of add for a stack
            } else {
                if (st.isEmpty() || st.pop() != bracket.get(c)) {
                    return false;
                }
            }
        }
        
        return st.isEmpty();
    }
}
